### PR TITLE
feat: Add OTLP handler to send events using OpenTelemetry

### LINKED ISSUE #121: Use OTLP instead of Libhoney

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
## Which problem is this PR solving?
Adds option of sending events using OpenTelemetry formatted data (OTLP). Includes a new config option that defaults to libhoney (current export format). 

- Closes #121 

## Short description of the changes
- Add new config option`HANDLER_TYPE` that defaults to `libhoney`
- Add new otel event handler that uses OTLP to export data, named `otel`
- Update k8s metadata lookup to return a map[string]string so it can be used by both handlers
- Updates existing tests to verify UIDs from k8s resources are strings

## How to verify that this has the expected result
No change by default. When setting the new env var to `otel`, events are sent to Honeycomb using OTLP instead of libhoney.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -149,6 +152,7 @@ func NewConfig() Config {
 		AdditionalAttributes:          utils.LookupEnvAsStringMap("ADDITIONAL_ATTRIBUTES"),
```
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply config/config.go:152]: I'm not sure we need to handle the OTel resource attributes env var, I think SDK should detect and apply them anyway.

File: handlers/event_handler.go [hunk 2]
```diff
@@ -13,3 +15,15 @@ type EventHandler interface {
 	Close()
 	handleEvent(event assemblers.Event)
 }
+
+// NewEventHandler returns an event handler based on the config's selected handler type.
+func NewEventHandler(config config.Config, cachedK8sClient *utils.CachedK8sClient, eventsChannel chan assemblers.Event, version string) EventHandler {
+	var eventHandler EventHandler
+	switch config.EventHandlerType {
+	case "libhoney":
```
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply handlers/event_handler.go:23]: I think it's valid if someone changed the handler type to an unknown value and the agent failed to start. I'm okay changing to move the default case to libhoney and log a warning though.

File: handlers/otel_handler.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+// NewOtelHandler creates a new event handler that sends events using OpenTelemetry
+func NewOtelHandler(config config.Config, k8sClient *utils.CachedK8sClient, eventsChan chan assemblers.Event, version string) EventHandler {
+	otelShutdown, err := otelconfig.ConfigureOpenTelemetry(
+		otelconfig.WithServiceName(config.Dataset),
+		otelconfig.WithServiceVersion(version),
+		otelconfig.WithExporterEndpoint(config.Endpoint),
+		otelconfig.WithHeaders(map[string]string{
+			"x-honeycomb-team":    config.APIKey,
+			"x-honeycomb-dataset": config.Dataset,
```
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply handlers/otel_handler.go:39]: The OTel SDK does enable metrics by default which I disabled for now. I think we're good to start with traces and re-add options to support other data signals if needed.

File: handlers/otel_handler.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	if event.Request() != nil {
+		span.SetAttributes(
+			semconv.HTTPMethod(event.Request().Method),
+			semconv.UserAgentOriginal(event.Request().Header.Get("User-Agent")),
+			semconv.HTTPRequestBodySize(int(event.Request().ContentLength)),
+		)
+		if handler.config.IncludeRequestURL {
+			url, err := url.ParseRequestURI(event.Request().RequestURI)
+			if err == nil {
+				span.SetAttributes(
```
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply handlers/otel_handler.go:140]: Done in https://github.com/honeycombio/honeycomb-network-agent/pull/290/commits/c5b6f854191395f1713c2669ea93223bb979be46

File: handlers/event_handler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply handlers/event_handler.go:24]: Done in https://github.com/honeycombio/honeycomb-network-agent/pull/290/commits/a7c5bb8077d612ac47b87e5093e8e118df0cf7e6

File: handlers/otel_handler.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+					semconv.HTTPURL(url.Path),
```
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply handlers/otel_handler.go]: Done in https://github.com/honeycombio/honeycomb-network-agent/pull/290/commits/0f8b7a7331227bef8071a41c6a44e233bc82a601

File: handlers/otel_handler.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+			semconv.ServerSocketAddress(event.DstIp()),
+		),
+		trace.WithAttributes(attrs...),
+	)
+	defer span.End(trace.WithTimestamp(endTime))
+
+	// request attributes
+	if event.Request() != nil {
+		span.SetAttributes(
+			semconv.HTTPMethod(event.Request().Method),
```
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply handlers/otel_handler.go:131]: There are some constant values for methods, but I think using the base key with a value get get from the request is fine.

I think using the same, newer version is good enough for now so at least libhoney and otel handler events are comparable. We can log a ticket to track duping attrs later 👍🏻

[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply handlers/otel_handler.go]: Update in https://github.com/honeycombio/honeycomb-network-agent/pull/290/commits/7cc7c6c2a9aea29b335af64fa73954390565430a

[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply handlers/otel_handler.go]: Created follow-up ticket to dupe attrs if needed: https://github.com/honeycombio/honeycomb-network-agent/issues/292

File: config/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply config/config.go:152]: I tested this and if you set `OTEL_RESOURCE_ATTRIBUTES` env var, the OTel SDK automatically picks them up.

My deployment yaml
image

honeycomb event:
image

File: handlers/otel_handler.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+	// response attributes
+	if event.Response() != nil {
+		span.SetAttributes(
+			semconv.HTTPResponseStatusCode(event.Response().StatusCode),
+			semconv.HTTPResponseBodySize(int(event.Response().ContentLength)),
+		)
+		// by this point, we've already extracted headers based on HTTP_HEADERS list
+		// so we can safely add the headers to the event
+		span.SetAttributes(headerToAttributes(false, event.Request().Header)...)
```
Comment (JamieDanielson) [inline handlers/otel_handler.go:159]:
```suggestion
		span.SetAttributes(headerToAttributes(false, event.Response().Header)...)
```

File: handlers/otel_handler.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+// headerToAttributes converts a http.Header into a slice of OpenTelemetry attributes
+func headerToAttributes(isRequest bool, header http.Header) []attribute.KeyValue {
+	var prefix string
+	if isRequest {
+		prefix = "http.request.header"
+	} else {
+		prefix = "http.response.header"
+	}
+	attrs := []attribute.KeyValue{}
+	for key, val := range header {
```
Comment (JamieDanielson) [inline handlers/otel_handler.go:243]: I was working on a test for this and ended up iterating through one more time to get clean strings for header values. I'm not sure if that's right but it did show up more cleanly as attribute values.

Also I adjusted to further match the recommendation of the semantic conventions, which is to lowercase the key and replace hyphens with underscores. 

What do you think? https://github.com/honeycombio/honeycomb-network-agent/commit/8537b56a8abca48f06c9d1d813619acdb5fcb7c4

**Review by JamieDanielson — COMMENTED** [review]
Small change needed in setting the response headers, and then a suggestion based on more closely following the semconv recommendation. I included those changes and some adjustments to the smoke test deployment in [another branch](https://github.com/honeycombio/honeycomb-network-agent/compare/mike/otel...jamie.http-headers-updates?expand=1), let me know what you think!

**Pull request merged**